working/0125-static-immutability/feature-specification.md

```diff
@@ -0,0 +1,230 @@
+# Shared immutable objects
+
+leafp@google.com
+
+Status: Draft
+
+This describes a possible solution for:
+ - [Communication between isolates](https://github.com/dart-lang/language/issues/124)
+ - [Building immutable collections](https://github.com/dart-lang/language/issues/117)
+ - [Unwanted mutation of lists in Flutter](https://github.com/dart-lang/sdk/issues/27755)
+
+## Summary
+
+This describes a way to declare classes that produce deeply immutable object
+graphs that are shared across isolates. 
+
+## Syntax
+
+We add a section to class headers for expressing class and generic constraints,
+along with an "immutable" constraint.
+
+```dart
+class Value<T> extends Scalar<T> implements Constant is immutable
+  where T is immutable {
+}
+```
+
+Mixin declarations may also be marked `immutable`.
+
+Generic method headers may also express generic constraints.
+
+```
+foo<S, T, where T is immutable>(Value<T> v) {
+
+}
+```
+
+### Alternative syntax 1
+
+Instead of adding constraints, a simpler approach is to add a marker interface
+`Immutable`.  The property expressed by the constraint `T is immutable` then
+becomes expressed by `implements Immutable` in the case of a class, or `T
+extends Immutable` in the case of a type variable `T`.
+
+### Alternative syntax 2
+
+Instead of adding general constraints, we could expose a dedicated syntax.  For
+example, this proposal from @yjbanov.
+
+```dart
+data Value<data T> extends Scalar<T> {
+}
+
+foo<S, data T>(Value<T> v) {
+}
+
+```
+
+
+## Static checking
+A class marked with `immutable` is subject to the following additional static
+checks.
+
+- Every field in an immutable class (including any superclass fields) must be
+  final.
+- Every field in an immutable class (including any superclass fields) must have
+  a static type which is immutable.
+- Every other class which implements the interface of an immutable class
+  (including via extension or mixing in) must also be immutable.
+
+The types `int`, `double`, `bool`, `String`, `Type`, and `Symbol` are considered
+immutable.
+
+## Generated methods
+
+We may wish to consider automatically generating hashCode and equality methods
```

Comment:
> This is still dangerous because auto-generated code can only rely on _fields_, not logical properties

I was a bit concerned about this.  @munificent seemed to feel it was a non-issue.  Something to think about.

> We cannot define an automatic equality based on _private_ fields

I don't understand why not?  Can you elaborate?

Reply:
Maybe you can, but since nobody knows the order or names of the private members, it is nigh impossible to make another implementation of the same interface that has the same equality/hash-code. So, rather than "we can't", it's probably "we shouldn't". It exposes the (arbitrary) internal implementation in a public method.

Reply:
I don't really understand this.  Suppose I have class.  It has private fields.  I want to put it in a hash table.  What is it that you suggest I do, if not hash the private fields?  Make the fields public so that I can hash them?  And if it's reasonable for me to do it, why is it not reasonable for a compiler to do it for me?

Reply:
True, it works. That class interface has private fields, so it is effectively unimplementable, so my concern that nobody else can implement the same equality is probably not a real problem.

My concern about hash code based on private fields may (I don't remember any more) have come from subclassing. If you subclass an immutable class with automatic hash codes and private fields in the superclass from another library, then you can't access the private field. You can call `super.hashCode`, and you might even be able to call `super.==`, so it could work.
```dart
immutable class C {
  final int _i;
  C(this._i);
  // generated:
  int get hashCode => hash1(_i.hashCode);
  bool operator==(Object other) => other is C && _i == other._i;
}
... other library ...
immutable class D {
  final int x;
  D(this.x, int i) : super(i);
  // generate what?
}
```
Then the hashCode/equality of `D` would have to be something like:
```dart
  int get hashCode => hash2(super.hashCode, x.hashCode);
  bool operator==(Object other) => other is D && x == other.x && super == other;
```
That could work.
It's still not symmetric, `C(1) == D(2, 1)` and `D(2, 1) != C(1)`. That's a separate, and general, problem with a equality and subclassing a concrete class. You should never subclass a concrete class, but I guess that rule is hard to sell :)